python/orca/tutorial/NCF/process_xshards.py

```diff
@@ -51,38 +53,86 @@ def ng_sampling(data, user_num, item_num):
     return data_XY
 
 
-def split_dataset(data):
+def split_dataset(data, total_cols):
+    data = data.loc[:, total_cols]
     train_data, test_data = train_test_split(data, test_size=0.2, random_state=100)
     return train_data, test_data
 
 
-def prepare_data(dataset_dir):
-    data = read_csv(os.path.join(dataset_dir, 'ratings.dat'),
-                    sep="::", header=None, names=['user', 'item'],
-                    usecols=[0, 1], dtype={0: np.int64, 1: np.int64})
-    data = data.partition_by("user")
-
-    user_set = set(data["user"].unique())
-    item_set = set(data["item"].unique())
+def prepare_data(dataset_dir, num_ng=4, total_cols=['user', 'item', "label"]):
+    # Load the train and test datasets
+    data_1 = read_csv(
+        os.path.join(dataset_dir, 'users.dat'),
+        sep="::", header=None, names=['user', 'gender', 'age', 'occupation', 'zipcode'],
+        usecols=[0, 1, 2, 3, 4])
+    data_2 = read_csv(
+        os.path.join(dataset_dir, 'ratings.dat'),
+        sep="::", header=None, names=['user', 'item'],
+        usecols=[0, 1])
+    data_3 = read_csv(
+        os.path.join(dataset_dir, 'movies.dat'),
+        sep="::", header=None, names=['item', 'category'],
+        usecols=[0, 2])
+
+    # calculate numbers of user and item
+    user_set = set(data_2["user"].unique())
+    item_set = set(data_2["item"].unique())
     user_num = max(user_set) + 1
     item_num = max(item_set) + 1
 
-    data = data.transform_shard(lambda shard: ng_sampling(shard, user_num, item_num))
-    train_data, test_data = data.transform_shard(split_dataset).split()
-    return train_data, test_data, user_num, item_num
+    # Negative sampling
+    data_2 = data_2.transform_shard(lambda shard: ng_sampling(shard, user_num, item_num, num_ng))
+
+    # Merge XShards
+    data = data_1.merge(data_2, on='user')
+    data = data.merge(data_3, on='item')
+    data = data.partition_by("user", data.num_partitions())
+
+    # Category encoding
+    strid = StringIndexer('gender')
+    data = strid.fit_transform(data)
+    strid.setInputCol('zipcode')
+    data = strid.fit_transform(data)
+    strid.setInputCol('category')
+    data = strid.fit_transform(data)
+
+    # Calculate num_embeddings for each categorical features
+    num_embed_cat_feats = []
+    num_embed_cat_feats.append(user_num)
+    num_embed_cat_feats.append(item_num)
+    cat_feat_set = set(data["gender"].unique())
+    num_embed_cat_feats.append(max(cat_feat_set)+1)
+    cat_feat_set = set(data["occupation"].unique())
+    num_embed_cat_feats.append(max(cat_feat_set)+1)
+    cat_feat_set = set(data["zipcode"].unique())
+    num_embed_cat_feats.append(max(cat_feat_set)+1)
+    cat_feat_set = set(data["category"].unique())
+    num_embed_cat_feats.append(max(cat_feat_set)+1)
+
+    # Split dataset
+    train_data, test_data = data.transform_shard(
+        lambda shard: split_dataset(shard, total_cols)).split()
+    return train_data, test_data, num_embed_cat_feats
```

Comment:
what about this method returns feature_cols and label_cols directly for train to use?